### PR TITLE
C2 mirna yacout halls of fame

### PR DESCRIPTION
**Pull request by Mirna-Yacout**
#198 #199

@AmrARaouf @karimngr @Omar-Hossam @Fattouh92

### CODE REVIEW

**Comment by Mirna-Yacout** [comment]
@AmrARaouf @Omar-Hossam please review code and documentation

**Comment by Mirna-Yacout** [comment]
@karimngr please review scenario

File: arability/app/models/authentication.rb [hunk 1]
```diff
@@ -55,30 +55,67 @@ def self.remove_conn(current_gamer_id, provider)
...
+  #  returns the list of common friends
   # Failure:
   #  returns nil if no authentication is found
-  def self.get_common_twitter_followers(current_gamer)
-    auth = Authentication.find_by_gamer_id_and_provider(current_gamer.id, "twitter")
-    if (auth.nil?)
-      return nil
+  def self.get_common_twitter_friends(current_gamer)
+    begin
+      auth = Authentication.find_by_gamer_id_and_provider(current_gamer.id, "twitter")
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: please make your line not exceed 80 characters

File: arability/app/models/authentication.rb [hunk 2]
```diff
@@ -55,30 +55,67 @@ def self.remove_conn(current_gamer_id, provider)
...
+      if (auth.nil?)
+        return nil
+      end
+      result = JSON.parse(open("https://api.twitter.com/1/friends/ids.json?user_id=#{auth.gid}").read)
+      followers = Array.new(result["ids"])
+      common = Array.new
+      i = 0
+      while i < followers.count
+        if Authentication.exists?(gid: followers.at(i), provider: "twitter")
+          common.push(Authentication.find_by_gid_and_provider(followers.at(i), "twitter").gamer_id)
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: please make you line not exceed 80 characters

File: arability/app/models/authentication.rb [hunk 3]
```diff
@@ -55,30 +55,67 @@ def self.remove_conn(current_gamer_id, provider)
...
+  #  This method is to retrieve the list of common arability friends and Facebook friends
+  # Parameters:
+  #  current_gamer: the record in Gamer table for the current user
+  # Success:
+  #  returns the list of common followers
+  # Failure:
+  #  returns nil if no gamer token is found
+  def self.get_common_facebook_friends(current_gamer)
+    begin
+      auth = Authentication.find_by_gamer_id_and_provider(current_gamer.id, "facebook")
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: please make your line not exceed 80 characters

File: arability/app/models/authentication.rb [hunk 4]
```diff
@@ -55,30 +55,67 @@ def self.remove_conn(current_gamer_id, provider)
...
+      if auth.nil?
+      return nil
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: indent line

File: arability/app/models/authentication.rb [hunk 5]
```diff
@@ -55,30 +55,67 @@ def self.remove_conn(current_gamer_id, provider)
...
+      end
+      @graph = Koala::Facebook::API.new(auth.token)
+      friends = @graph.get_connections("me", "friends")
+      common = Array.new
+      i = 0
+      while i < friends.count
+        if Authentication.exists?(gid: friends.at(i)["id"], provider: "facebook")
+          common.push(Authentication.find_by_gid_and_provider(friends.at(i)["id"], "facebook").id)
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: please reduce the previous 2 lines to not exceed 80 characters, and if you want be able to do that, reduce their size as much as you can

File: arability/app/models/gamer.rb [hunk 6]
```diff
@@ -227,35 +227,108 @@ def won_prizes?(score, level)
...
+#     Creates a new gamer using the email and password
+def self.find_for_google_oauth2(access_token, signed_in_resource=nil)
+    data = access_token.info
+    gamer = Gamer.where(:email => data["email"]).first
+
+    unless gamer
+         gamer = Gamer.create(
+              email: data["email"],
+              password: Devise.friendly_token[0,20]
+             )
```
Comment (omar-mohyeldin) [inline arability/app/models/gamer.rb]: please adjust the previous code's indentation, and make it follow the code conventions

File: arability/app/models/gamer.rb [hunk 7]
```diff
@@ -227,35 +227,108 @@ def won_prizes?(score, level)
...
+  current_gamer_rank = nil
+  rank = 1
+  gamers.each do |user|
+    if user == current_gamer
+      current_gamer_rank = rank
+    end
+  rank = rank + 1
+  end
+  return current_gamer_rank
+end
```
Comment (omar-mohyeldin) [inline arability/app/models/gamer.rb]: please adjust all of the previous codes' indentation

File: arability/app/views/games/_halloffame.html.erb [hunk 8]
```diff
@@ -0,0 +1,186 @@
+<script type="text/javascript" charset="utf-8">
+
```
Comment (omar-mohyeldin) [inline arability/app/views/games/_halloffame.html.erb]: remove spaces from empty line

File: arability/app/views/games/_halloffame.html.erb [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        <td><%= user.highest_score %></td>
+      </tr>
+      <br>
+      <% end %>
+    </tbody>
+  </table>
+  <div align="center" style="color:white;"><%= paginate @gamers %></div>
+  <% end %>
+  <% end %>
+</div>
```
Comment (omar-mohyeldin) [inline arability/app/views/games/_halloffame.html.erb:187]: please adjust the previous code's indentation with if conditions

File: arability/app/views/games/fame.js.erb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+$('.zone').empty();
+$('.zone').append('<%= escape_javascript(render("halloffame")) %>');
```
Comment (omar-mohyeldin) [inline arability/app/views/games/fame.js.erb]: please use double quotations for ".zone"

File: arability/app/views/games/game.html.erb [hunk 11]
```diff
@@ -206,20 +206,30 @@ function setButtons(){
...
-    <% if I18n.locale==:en %>
-    $('.zone').append('<div id = "loading-server"><p style = " position:absolute; margin-top: 150px; margin-left: 150px; margin-right: 150px; font-size: 1000%; color: white; ">Loading</p></div>');
-    <% else I18n.locale==:ar %>
-    $('.zone').append('<div id = "loading-server"><p style = " position:absolute; margin-top: 150px; margin-left: 150px; margin-right: 150px; font-size: 1000%; color: white; ">تحميل</p></div>');
-    <% end %>
-    $('#loading-server').fadeTo(0,0);
-    $('#loading-server').fadeTo(1000,1);
-    $('#loading-server').fadeTo(1000,0);
-    $.get("/games/halloffame");
+    $('.zone').append('<div id="loading-server" class="progress"><div class="bar" style="width: 60%;"></div></div>');
```
Comment (omar-mohyeldin) [inline arability/app/views/games/game.html.erb]: use double quotations for ".zone"

File: arability/app/views/games/game.html.erb [hunk 12]
```diff
@@ -206,20 +206,30 @@ function setButtons(){
...
+    $.get("/games/fame");
+    $(window).load(function(){
+      $('#loading-server').fadeOut();
```
Comment (omar-mohyeldin) [inline arability/app/views/games/game.html.erb]: use double quotations

**Comment by omar-mohyeldin** [comment]
@Mirna-Yacout please modify your code

File: arability/app/controllers/games_controller.rb [hunk 13]
```diff
@@ -262,7 +262,6 @@ def record_synonym
       keyword_id: params[:keyword_id]).first 
   end
 
-  # Author:
```
Comment (AmrARaouf) [inline arability/app/controllers/games_controller.rb:274]: ? why?

File: arability/app/models/authentication.rb [hunk 14]
```diff
@@ -55,30 +55,67 @@ def self.remove_conn(current_gamer_id, provider)
...
-    followers = Array.new(result["ids"])
-    common = Array.new
-    i = 0
-    while i<followers.count
-      if Authentication.exists?(gid: followers.at(i), provider: "twitter")
-        common.push(Authentication.find_by_gid_and_provider(followers.at(i), "twitter").gamer_id)
-        common.push(current_gamer.id)
+  end
+
+  # Author:
```
Comment (AmrARaouf) [inline arability/app/models/authentication.rb:90]: also indent this doc properly

File: arability/app/models/gamer.rb (hunk 6 above)
Comment (AmrARaouf) [inline arability/app/models/gamer.rb]: this method should be deleted. someone added it again

File: arability/app/models/gamer.rb [hunk 15]
```diff
@@ -227,35 +227,108 @@ def won_prizes?(score, level)
...
+    unless gamer
+         gamer = Gamer.create(
+              email: data["email"],
+              password: Devise.friendly_token[0,20]
+             )
     end
+    gamer
+end
+
+# Author:
```
Comment (AmrARaouf) [inline arability/app/models/gamer.rb]: indent properly please, 1 space for the headers and 3 for the body

File: arability/app/models/gamer.rb [hunk 16]
```diff
@@ -227,35 +227,108 @@ def won_prizes?(score, level)
...
+  gamers.each do |user|
+   if user == current_gamer
+      current_gamer_rank = rank
+   end
+   rank = rank + 1
+  end
+  return current_gamer_rank
+end
+
+# Author:
```
Comment (AmrARaouf) [inline arability/app/models/gamer.rb]: and again

File: arability/app/models/gamer.rb [hunk 17]
```diff
@@ -227,35 +227,108 @@ def won_prizes?(score, level)
...
+  gamers.each do |user|
+    if user == current_gamer
+      current_gamer_rank = rank
+    end
+    rank = rank + 1
+  end
+  return current_gamer_rank
+end
+
+# Author:
```
Comment (AmrARaouf) [inline arability/app/models/gamer.rb]: and again

File: arability/app/views/games/game.html.erb [hunk 18]
```diff
@@ -206,20 +206,30 @@ function setButtons(){
...
+  //   Mirna Yacout
+  // Description:
+  //   views the hall of fame of all arability gamers showing their last 3 trophies, rank and highest score
+  //   giving the user the option to filter with his facebook or twitter friends and view his rank
+  // Parameters:
+  //   None.
+  // Success:
+  //   view the hall of fame successfully
+  // Failure:
+  //   internet connection
```
Comment (AmrARaouf) [inline arability/app/views/games/game.html.erb]: you should say what happens in case of an internet connection problem

**Comment by AmrARaouf** [comment]
tests please

**Comment by Mirna-Yacout** [comment]
@AmrARaouf review tests

File: arability/app/views/games/_halloffame.html.erb [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+      <tr>
+        <td><%= t(:rank) %></td>
+        <td><%= t(:name) %></td>
+        <td><%= t(:highest_score) %></td>
+        <td><%= t(:trophies) %></td>
+      </tr>
+    </thead>
+    <tbody align="center">
+      <% @gamers = Gamer.order("highest_score DESC").page(params[:page]).per(5) %>
+      <% @gamers.each do |user| %>
```
Comment (omar-mohyeldin) [inline arability/app/views/games/_halloffame.html.erb:73]: please indent following code to be indent by 2 extra spaces than the "do" statement

File: arability/app/views/games/_halloffame.html.erb [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+        </thead>
+        <tbody align="center">
+          <% @gamers = Gamer.where(id: common).order("highest_score DESC").page(params[:page]).per(5) %>
+          <% @gamers.each do |user| %>
+            <% if user == current_gamer %>
+              <tr style="background-color: #DDDDDD">
+           <% else %>
+              <tr>
+           <% end %>
+              <td><%= rank = Gamer.get_facebook_rank(user) %></td>
```
Comment (omar-mohyeldin) [inline arability/app/views/games/_halloffame.html.erb:176]: please remove 2 spaces from the indentation of the following code as it should be alligned with the "if", "else" and end statements

File: arability/app/views/games/_halloffame.html.erb [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+          </tr>
+        </thead>
+        <tbody align="center">
+          <% @gamers = Gamer.where(id: common).order("highest_score DESC").page(params[:page]).per(5) %>
+          <% @gamers.each do |user| %>
+            <% if user == current_gamer %>
+              <tr style="background-color: #DDDDDD">
+           <% else %>
+              <tr>
+           <% end %>
```
Comment (omar-mohyeldin) [inline arability/app/views/games/_halloffame.html.erb]: "else" and "end" should be alligned with "if"

File: arability/spec/models/get_ranks_spec.rb [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    current_gamer
+    rank = Gamer.get_facebook_rank(current_gamer)
+    rank.should eq(nil)
+  end
+  it "should return gamer rank within twitter friends registered in arability" do
+    current_gamer
+    rank = Gamer.get_twitter_rank(current_gamer)
+    rank.should eq(nil)
+  end
+end
```
Comment (omar-mohyeldin) [inline arability/spec/models/get_ranks_spec.rb:37]: please leave an empty line between each 2 tests

File: arability/spec/models/get_social_friends_spec.rb [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+  #   Koala::Facebook::API.stub(:new).with(fbauthentication.token) { raise Faraday::Error::ConnectionFailed }
+  #   Authentication.get_common_facebook_friends(current_gamer).should be_false
+  # end
+  # it "should retrieve list of facebook friends" do
+  #   current_gamer
+  #   fbauthentication
+  #   result = Authentication.get_common_facebook_friends(current_gamer)
+  #   expect(result).to eq([1])
+  # end
+end
```
Comment (omar-mohyeldin) [inline arability/spec/models/get_social_friends_spec.rb:69]: please leave an empty line between each 2 tests, and remove commented lines

**Comment by omar-mohyeldin** [comment]
@Mirna-Yacout please modify your code

File: arability/app/views/games/_halloffame.html.erb (hunk 20 above)
Comment (Mirna-Yacout) [inline arability/app/views/games/_halloffame.html.erb:176]: yes but it should be indented because of the <tr> tag

Comment (omar-mohyeldin) [inline arability/app/views/games/_halloffame.html.erb:176]: which tag?

Comment (Mirna-Yacout) [inline arability/app/views/games/_halloffame.html.erb:176]: inside the if condition

Comment (omar-mohyeldin) [inline arability/app/views/games/_halloffame.html.erb:176]: yes, but I'm talking about the lines after the end of the if condition

**Comment by omar-mohyeldin** [comment]
@Mirna-Yacout code verified

**Comment by abulazm22** [comment]
Verified

**Comment by ghost** [comment]
it is taking too long to open the page please fix it

File: arability/app/models/authentication.rb [hunk 24]
```diff
@@ -55,30 +55,71 @@ def self.remove_conn(current_gamer_id, provider)
   # Author:
   #  Mirna Yacout
```
Comment (AmrARaouf) [inline arability/app/models/authentication.rb]: please modify this doc's indentation

**Comment by AmrARaouf** [comment]
please pull master

**Comment by Mirna-Yacout** [comment]
@karimngr @AmrARaouf check now

**Pull request closed without merging**